add bluetooth support for seika notetaker

feerrenrut opened the pull request:
### Link to issue number:
fixes https://github.com/nvaccess/nvda/issues/13142

### Summary of the issue:
Seika Notetaker can not be autodetected via bluetooth

### Description of how this pull request fixes the issue:
Add a way to identify the Seika device when connected by bluetooth

As per https://github.com/nvaccess/nvda/issues/13142#issuecomment-996441860 the bluetooth name for Seika devices should be `TSM`, a space, then 4 digits.

### Testing strategy:
Ask Seika users to test when connected via bluetooth.

### Known issues with pull request:
If ALL Seika devices use this bluetooth name, we may not be able to differentiate which driver to use `seika` vs `seika notetaker`.

### Change log entries:
```
New features
- Seika Notetaker can now be auto-detected when connected via Bluetooth
```

### Code Review Checklist:
- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

feerrenrut committed:
add bluetooth support for seika notetaker

moyanming commented:
Hi, @feerrenrut 
The Bluetooth name of the Seika is "TS5 abcd", where "abcd" is a four-digit number.
Seika: TS5, a space, and a four-digit number.
Seika Notetaker: TSM, a space, and a four-digit number.

I can test both Seika and Seika Notetaker Braille display.

Best Regards,

cary-rowen commented:
@feerrenrut 
I tested the build Version, and unfortunately, NVDA did not connect to my device via Bluetooth.
Shows an error dialog
The following is the log with HWIO enabled:
[log_#13191.txt](https://github.com/nvaccess/nvda/files/7772656/log_.13191.txt)

feerrenrut committed:
attempt to connect to seikantk using serial over bluetooth

feerrenrut committed:
fix error in seika driver

feerrenrut committed:
byte ordering isn't required for a single byte

Elsewhere the bytes are in 'big' byte ordering

cary-rowen commented:
Hello, I tested the latest build version, NVDA seems to behave the same as last time, and it still cannot connect to Seika Notetaker via Bluetooth. Here is the log:

[log.txt](https://github.com/nvaccess/nvda/files/7829163/log.txt)

feerrenrut committed:
fix lint check

feerrenrut committed:
fix portType vs match.port mixup

cary-rowen commented:
There is a short wait time when connecting, and the final result is that the connection fails.
[log-01-09.txt](https://github.com/nvaccess/nvda/files/7834425/log-01-09.txt)

feerrenrut committed:
fix lint

feerrenrut committed:
split command collection approaches

HID and Serial seem to require different approaches

feerrenrut commented:
@cary-rowen Thank you for testing and providing the log. I have had to guess about the issue here, could you please test both bluetooth and USB and provide a log for me?

With HID the first and third byte is discarded, for the serial connection the bytes coming through look like the standard command. In the log provided the following was received: `\xff\xff\xa2\x0b\x16\x18\x18TSM V1.0`. This looks like:
- Command: `\xff\xff\xa2` = `SEIKA_INFO`
- argument length: `x0b` = 11
- number of buttons: `\x16` = 22
- number of cells: `\x18` = 24
- number of routing keys: `\x18` = 24
- description: "TSM V1.0"

cary-rowen commented:
Hi @feerrenrut , the latest build has been able to successfully connect devices using bluetooth.
But there are multiple optional items in the "Port" combo box, one of which is 'friendlyName': '蓝牙链接上的标准串行 (COM4)'
NVDA freezes when this is checked and the "OK" button is pressed,
The corresponding entry in the log for this item is as follows:
{'hardwareID': 'BTHENUM\\{00001101-0000-1000-8000-00805f9b34fb}_LOCALMFG&0000', 'port': 'COM4', 'bluetoothAddress': 0, 'friendlyName': '蓝牙链接上的标准串行 (COM4)'}

That is, there are a total of four items in the Port combo box:
1. Automatic
2. Bluetooth
3. Bluetooth Serial: COM3 (TSM 3366)
4. Serial: 蓝牙链接上的标准串行 (COM4)
The first three items can be successfully connected to Seika Notetaker, and the fourth item causes NVDA to freeze.

Grateful

moyanming commented:
> 

Hi @feerrenrut 
The Seika Notetaker sending back the correct data means the connection is successfully now.
\xff\xff\xa2\x0b\x16\x18\x18TSM V1.0

BTW, both USB-HID and Bluetooth Serial COM port use the same data protocol.

moyanming commented:
> 

Hi @cary-rowen 
There have two Bluetooth UUIDs, one for Tx and another for Rx, and this is why there are two Bluetooth Serial Com ports.
I think the Bluetooth Serail Com port which has "TSM abcd" is the right one for the communication.

feerrenrut commented:
@moyanming can you test the [latest PR build](https://ci.appveyor.com/api/buildjobs/ty1n9ad9jr9927vv/artifacts/output%2Fnvda_snapshot_pr13191-24426%2C82c1e231.exe) with bluetooth auto detection and report your findings. I believe this should be working, if so I'd like to merge this PR. Then we can raise a new issue and look at the listing of ports and see if we can prevent the wrong one from being listed.

cary-rowen commented:
And, another question, in the case of a successful connection, the user re-selects the same port in the combo box and clicks the "OK" button, then what is our expected performance? Currently makes Seika Notetaker show nothing as if it's disconnected.

moyanming commented:
> @moyanming can you test the [latest PR build](https://ci.appveyor.com/api/buildjobs/ty1n9ad9jr9927vv/artifacts/output%2Fnvda_snapshot_pr13191-24426%2C82c1e231.exe) with bluetooth auto detection and report your findings. I believe this should be working, if so I'd like to merge this PR. Then we can raise a new issue and look at the listing of ports and see if we can prevent the wrong one from being listed.

Hi, @feerrenrut 
I have tested the PR build and used the USB and BT respectively, I also test the Automatic which is not select the specific Braille display.
The Port list of the Seika Notetatker is: Automatic, USB, Bluetooth, Serial (com1), Bluetooth Serial with TSM 2002, and Serial without TSM abcd.
The following log files all work fine:
[BT_SeikaNotetaker_COM port with 'TSM abcd'_OK.log](https://github.com/nvaccess/nvda/files/7837769/BT_SeikaNotetaker_COM.port.with.TSM.abcd._OK.log)
[USB_Automatic Braille display_OK.log](https://github.com/nvaccess/nvda/files/7837770/USB_Automatic.Braille.display_OK.log)
[USB_SeikaNotetaker_Automatic_OK.log](https://github.com/nvaccess/nvda/files/7837771/USB_SeikaNotetaker_Automatic_OK.log)
[USB_SeikaNotetaker_USB_OK.log](https://github.com/nvaccess/nvda/files/7837773/USB_SeikaNotetaker_USB_OK.log)
[BT_Automatic Braille display_OK.log](https://github.com/nvaccess/nvda/files/7837775/BT_Automatic.Braille.display_OK.log)
[BT_SeikaNotetaker_Automatic_OK.log](https://github.com/nvaccess/nvda/files/7837776/BT_SeikaNotetaker_Automatic_OK.log)
[BT_SeikaNotetaker_Bluetooth_OK.log](https://github.com/nvaccess/nvda/files/7837777/BT_SeikaNotetaker_Bluetooth_OK.log)

I got a crash issue when selecting the Bluetooth Serial COM port without TSM abcd:
[BT_SeikaNotetaker_COM port without 'TSM abcd' will crash.log](https://github.com/nvaccess/nvda/files/7837784/BT_SeikaNotetaker_COM.port.without.TSM.abcd.will.crash.log)

BTW, accidentally disconnecting the USB cable and then reconnecting the USB, the NVDA still can't connect automatically but needs to restart NVDA, here is the log:
[Accidentally disconnect-USB needs restart NVDA_issue.log](https://github.com/nvaccess/nvda/files/7837788/Accidentally.disconnect-USB.needs.restart.NVDA_issue.log)

feerrenrut committed:
Filter manual ports

Use the bluetoothName to filter manual ports,
matching the same approach used by auto detection
of the seika notetaker device.

In order to prevent a circular import,
the initialization of the detection data (and
new import of seika notetaker module method
was moved into a contained method.

feerrenrut committed:
fix seikantk tests

Add tests for serial (bluetooth)

feerrenrut commented:
Can you try the [latest build](https://ci.appveyor.com/api/buildjobs/wq80bk6c1o4k9g3p/artifacts/output%2Fnvda_snapshot_pr13191-24463%2C7a9dfdce.exe)
the Bluetooth Serial COM port without TSM abcd should no longer be listed.

cary-rowen commented:
Hi, the build seems to introduce a regression,:
Launching this version of NVDA for the first time, but the Seaka device is not turned on, navigating to "Seika Notetaker" in the braille display combo box NVDA throws an error, the log is as follows:
[Device not open, navigate in device selection combo box.txt](https://github.com/nvaccess/nvda/files/7875450/Device.not.open.navigate.in.device.selection.combo.box.txt)
The Seaka device enters Bluetooth pairing mode, and then starts this version of NVDA. At this time, Seaka can successfully display content through "Auto Connect", but the device selection dialog cannot be opened:
[Unable to open device selection dialog after connecting.txt](https://github.com/nvaccess/nvda/files/7875455/Unable.to.open.device.selection.dialog.after.connecting.txt)

lukaszgo1 commented on source/brailleDisplayDrivers/seikantk.py:
```diff
@@ -57,66 +59,121 @@
+		return braille.getSerialPorts(isSeikaBluetoothDeviceInfo)
 
-	def __init__(self, port="hid"):
+	def __init__(self, port=bdDetect.KEY_HID):
```
While I'm not very familiar with NVDA's braille display drivers, this is the only one which specified `HID` as the value for the `port` keyword argument in the constructor. For all others we either leave it with no default value in case of displays where port needs to be specified by the user, or for displays which can be auto-detected we assign `"auto"` to it. Could you add a comment here to explain why this driver is different?

feerrenrut replied:
The short answer is I was trying to avoid making changes if not necessary. Looking at this closer, it seems like it should never be "hid". I have tried to improve the docs and typing elsewhere. One confusing thing in the NVDA braille code is that "port" is used to refer to many different things, in different parts of the code it may be a COM port string, a string description of where to look (auto, usb, bluetooth), a HID device path string, or a DeviceMatch class instance.

feerrenrut committed:
seikantk init remove default val for port

There does not seem to be any situation where the port should be hid.

feerrenrut committed:
fix missing key during bluetooth detection

feerrenrut commented:
Thanks for testing again @cary-rowen [the latest build](https://ci.appveyor.com/api/buildjobs/toya6566jk9nkdaf/artifacts/output%2Fnvda_snapshot_pr13191-24464%2Ca1cd6508.exe) should address this.

cary-rowen commented:
@feerrenrut 
Yes, I tested this build and it works fine.

As mentioned, another question, in the case of a successful connection, the user re-selects the same port in the combo box and clicks the "OK" button, then what is our expected performance? Currently makes Seika Notetaker show nothing as if it's disconnected.
We may need to reopen an issue for this problem, as the problem is also related to the USB connection.
what do you think?

Thanks again for your efforts

feerrenrut commented:
@cary-rowen Let's get this change merged first, then come back and look at that issue. Do you have any other braille device you could test with? It would be helpful to know if it is specific to this driver or affects others too.

cary-rowen commented:
@feerrenrut 
Hi,
OK, but I currently only have Seaka devices to test.
I'm also hoping someone else can provide different feedback

seanbudd reviewed (commented):
Code changes and unit test coverage look good to me.
Can documentation and testing be ticked off yet?

feerrenrut commented:
Thanks @cary-rowen, can you raise a new issue for that please?

feerrenrut commented on source/brailleDisplayDrivers/seikantk.py:
```diff
@@ -57,66 +59,127 @@
+						bytesize=serial.EIGHTBITS,
+						stopbits=serial.STOPBITS_ONE,
+					)
+					# Does the device need to be sent SEIKA_CMD_ON as per USB-HID?
```
I think the answer is that it isn't required, I'd like to leave some comment here to highlight the difference between serial and HID.  But I'll convert it to a statement rather than a question.

feerrenrut committed:
update comments for SEIKA_CMD_ON

feerrenrut committed:
use US spelling for initialize

feerrenrut committed:
Update user docs for braille notetaker

cary-rowen commented:
Hi @feerrenrut,

I have open a new Issue #13243 to clarify the problem I said above.

Thanks

feerrenrut commented:
@cary-rowen / @moyanming or any other Sieka braille device user.

Could you verify the changes I have made to the userguide for the Seika Notetaker:
- [direct link to user guide](https://ci.appveyor.com/api/buildjobs/kntp5pk5xmkmmt6f/artifacts/output%2Fnvda_snapshot_pr13191-24474%2C4cc40b29_changes.html)
- [NVDA installer, with updated userguide](https://ci.appveyor.com/api/buildjobs/kntp5pk5xmkmmt6f/artifacts/output%2Fnvda_snapshot_pr13191-24474%2C4cc40b29.exe)

feerrenrut commented:
I want to verify that `d7` makes sense, I could not find anything about it in the Notetaker user guide.

cary-rowen commented:
Hi, @feerrenrut 

I use NVDA+Ctrl+F to find "Seika Notetaker".
There is only one result, the statement about auto-connect, which I think is correct, the device supports auto-connect.
Also, what do you mean by "D7", I don't understand.

all the best

moyanming commented:
> I want to verify that `d7` makes sense, I could not find anything about it in the Notetaker user guide.

Hi, @feerrenrut 
Do you mean the Dot-7 key on the Seika Notetaker? BTW, there are 8 dots keys on the Seika Notetaker top panel, and from left to right is Dot-7, Dot-3, Dot-2, Dot-1, Dot-4, Dot-5, Dot-6, Dot-8.
The single pressed Dot-7 key is always defined as Backspace, the single pressed Dot-8 key is always defined as Enter. Dot1, 2, 3, 4, 5, 6 makes a 6-dot Braille table, with Dot7 and Dot8 there is an 8-dot Braille table. In the 8-dot Braille table, the Dot-7 key adding to the lower case letter will get an uppercase letter, e.g. dot15 is letter 'e', but dot157 is letter 'E'.

moyanming commented:
There is no uppercase letter within the 6-dot Braille table, UEB, etc. and of course, there has a more complicated rule to define the meaning of the uppercase letter and other symbols, e.g. there should make sure the "abc", "Abc", and "AbC" are clear in those tables.

But with Dot7 and Dot8 keys added on the Seika Notetaker, there is no problem and no need to remember the huge complicated rules when using the 8-dot Braille table.
This may be done by the Liblouis project.

feerrenrut commented:
@moyanming I wrote this table based on the gestures in the seikantk driver. The `d7` entry is this one:
`"kb:backspace": ("br(seikantk):d7",),`. Other drivers use `dot7` to to refer to dots. Can you confirm that `dot7` can be used to backspace characters?

feerrenrut commented:
@moyanming Is it possible to confirm how this command / gesture shows up in the input gestures dialog. I want to make sure this is consistent with other braille drivers.

moyanming commented:
> @moyanming I wrote this table based on the gestures in the seikantk driver. The `d7` entry is this one: `"kb:backspace": ("br(seikantk):d7",),`. Other drivers use `dot7` to to refer to dots. Can you confirm that `dot7` can be used to backspace characters?

Hi @feerrenrut 
For the code: there is a "_dotNames" that defined the d7 (means "dot7") and other dot keys, there should be changed everywhere if needed.
For the User guider: it is fine to use "dot7" to represent the dot 7 key of the Seika Notetaker.
Thanks!

feerrenrut commented:
Looking at the other drivers, the `hims`, `papenmeier`, and `brailleNote` drivers use `dx` eg `d7` as well. I'll update the user guide to refer to this as "dot 7"

feerrenrut committed:
use dot not d

For seika notetaker, userguide should refer to dot7 and dot8 not d7 and d8

cary-rowen commented:
@moyanming

Hi, There are currently two options with the word "Seika" in the braille device selection combo box, we don't seem to have reflected their differences in the user guide, can you try to mention this in the user guide, or do we The options that can be in the combo box themselves reflect their range of support, sorry this may be off topic. If you want, we can contact privately, my email is: 
manchen_0528@outlook.com

Thanks

feerrenrut commented:
@cary-rowen can you get a screenshot of this, or set NVDA to English and copy the speech output from the speech viewer. I want to make sure I understand the request.

I think there will be one entry for "seika" and one for "seika notetaker"?

feerrenrut commented:
Apologies, the last time I asked to check the user guide I linked to the changes file, I have corrected that link.

Can @cary-rowen and @moyanming please confirm the accuracy of the user guide:
- [direct link to html file](https://ci.appveyor.com/api/buildjobs/iysk8ceak02fjxuf/artifacts/output%2FuserGuide.html)
- [NVDA installer (also has the user guide) ](https://ci.appveyor.com/api/buildjobs/iysk8ceak02fjxuf/artifacts/output%2Fnvda_snapshot_pr13191-24483%2C82df1d26.exe)

moyanming commented:
> @moyanming
> 
> Hi, There are currently two options with the word "Seika" in the braille device selection combo box, we don't seem to have reflected their differences in the user guide, can you try to mention this in the user guide, or do we The options that can be in the combo box themselves reflect their range of support, sorry this may be off topic. If you want, we can contact privately, my email is: [manchen_0528@outlook.com](mailto:manchen_0528@outlook.com)
> 
> Thanks

Hi @cary-rowen 
For historical reasons, there is a Seika representing Seika v3/5/80 Braille display.
There is a new device type named Seika Notetaker Braille display published later and uses a different protocol compared to Seika Braille display.
We recommend using two different items for easy adding/modifying the driver.
BTW, you can select "Automatic" when connecting to Seika Notetaker now and this will work fine.

moyanming commented:
> 

Hi @feerrenrut 
Yes, there are two entry items: "Seika" and "Seika Notetaker".

moyanming commented:
Hi, @feerrenrut 
There are sub-type of Seika Notetaker: 16 cells Seika Notetaker, 24 cells Seika Notetaker, and 40 cells Seika Notetaker. There is no need to install the driver when using Seika Notetaker because those devices use USB-HID.

So please change from:
15.11. Seika Braille Displays
The Seika Version 3, 4 and 5 (40 cells), Seika80 (80 cells), and Notetaker (16 cells) braille displays from Nippon Telesoft are supported. You can find more information about the displays and the required drivers at https://en.seika-braille.com/down/index.html You must first install the USB drivers provided by the manufacturer.

Should change to:
15.11. Seika Braille Displays
The Seika Version 3, 4, and 5 (40 cells), Seika80 (80 cells), and Notetaker **(16, 24, 40 cells)** braille displays from Nippon Telesoft are supported. You can find more information about the displays and the required drivers at https://en.seika-braille.com/down/index.html. You must first install the USB drivers **when using Seika v3/4/5/80** provided by the manufacturer. **There is no need to install the driver when using Seika Notetaker.**

cary-rowen commented:
@moyanming @feerrenrut 
Ok, I see, but I understand that seika also has a V6Pro model in China, do you know about it? Is it possible to mention the device in the user guide as well?

Grateful

moyanming commented:
> @moyanming @feerrenrut Ok, I see, but I understand that seika also has a V6Pro model in China, do you know about it? Is it possible to mention the device in the user guide as well?
> 
> Grateful

Hi @cary-rowen 
Seika V6Pro is kind of Seika Notetaker has 40 cells.
The Seika Notetaker device name may vary but those devices should use "Seika notetaker" or "Automatic" compare to Seika v3/4/5/80 devices.

feerrenrut committed:
update seika user docs

feerrenrut commented:
I have split the information for the two different drivers. @moyanming Can you review this?

[installer](https://ci.appveyor.com/api/buildjobs/e9qr00wkbgy7vjm7/artifacts/output%2Fnvda_snapshot_pr13191-24484%2C200482ea.exe)
[direct user guide html](https://ci.appveyor.com/api/buildjobs/e9qr00wkbgy7vjm7/artifacts/output%2FuserGuide.html)

moyanming commented:
> 

Hi @feerrenrut 
That's great.

V6 and V6Pro are 40 cells Seika Notetaker, so please add V6 as well.
BTW, V6Pro has TTS voice supported but V6 hasn't.

So, please change:
15.11.2. MiniSeika (16, 24 cells), V6Pro (40 cells)
to:
15.11.2. MiniSeika (16, 24 cells), V6 and V6Pro (40 cells)

feerrenrut committed:
Include V6 seika device

feerrenrut commented:
@moyanming done. The build will complete shortly.

feerrenrut committed:
Update changes file for PR #13191

feerrenrut committed:
fix changes doc

moyanming commented:
> As per [#13142 (comment)](https://github.com/nvaccess/nvda/issues/13142#issuecomment-996441860) the bluetooth name for Seika devices should be `TSM`, a space, then 4 digits.
> 
> ### Testing strategy:
> Ask Seika users to test when connected via bluetooth.
> 
> ### Known issues with pull request:
> If ALL Seika devices use this bluetooth name, we may not be able to differentiate which driver to use `seika` vs `seika notetaker`.
> 

Hi @feerrenrut 
Thanks for your great work and the "Seika Notetaker" already supported by Bluetooth connection.

But the "Seika"( means "Seika v3/v5/v80") still can't connect with NVDA by using Bluetooth COM port, which is no the same type as "Seika Notetaker".
The Seika uses Bluetooth name: "TS5 abcd", but Seika Notetaker uses: "TSM abcd", where "abcd" are four digital numbers.
Do I add a new Issue about this?

Best Regards

cary-rowen commented:
@moyanming 
Can you create a new question for this? NVAccess cannot triage closed issues.
Also, I would like to get in touch with you privately, my email is manchen 0528@outlook.com

moyanming commented:
> 

Hi @cary-rowen 
I have report an issue #14242 and send to your email manchen_0528@outlook.com and please check you Inbox.
